### PR TITLE
test: run promtool test without docker on focal

### PR DESCRIPTION
**Pull request by tchaikov**
see also https://tracker.ceph.com/issues/49653

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by epuertat — APPROVED** [review]
Looks really good, @tchaikov! Thank you very much for taking care of this.

**Comment by tchaikov** [comment]
jenkins test make check

File: src/test/run-promtool-unittests.sh [hunk 1]
```diff
@@ -5,7 +5,11 @@ set -ex
...
 
-sudo docker run --rm \
+
+podman_or_docker=$1
+shift
+
+sudo ${podman_or_docker} run --rm \
          -v "$CEPH_ROOT":/ceph \
          --name=promtool \
          --network=host \
```
Comment (badone) [inline src/test/run-promtool-unittests.sh:15]: Do we need to specify a network?

Comment (badone) [reply src/test/run-promtool-unittests.sh:15]: I wonder whether we should do a "podman/docker rmi dnanexus/promtool:2.9.2" so we don't leave an unwanted container around?

File: src/test/CMakeLists.txt [hunk 2]
```diff
@@ -590,7 +590,32 @@ add_ceph_test(run-cli-tests ${CMAKE_CURRENT_SOURCE_DIR}/run-cli-tests)
...
+  execute_process(
+    COMMAND ${PROMTOOL_EXECUTABLE} test rules /dev/null
+    RESULT_VARIABLE rc
+    OUTPUT_QUIET)
+  if(NOT rc)
+    add_ceph_test(run-promtool-unittests
+      ${PROMTOOL_EXECUTABLE} test rules ${CMAKE_SOURCE_DIR}/monitoring/prometheus/alerts/test_alerts.yml)
+  else()
+    message(WARNING "'${PROMTOOL_EXECUTABLE} test rules' does not work, "
+      "please use a newer prometheus")
```
Comment (badone) [inline src/test/CMakeLists.txt:604]: We *could* fall back to the binaries from https://github.com/prometheus/prometheus/releases here if you are not averse to that suggestion?

File: src/test/CMakeLists.txt [hunk 3]
```diff
@@ -590,7 +590,32 @@ add_ceph_test(run-cli-tests ${CMAKE_CURRENT_SOURCE_DIR}/run-cli-tests)
...
+  if(PODMAN_EXECUTABLE)
+    add_ceph_test(run-promtool-unittests
+      {CMAKE_CURRENT_SOURCE_DIR}/run-promtool-unittests.sh "${PODMAN_EXECUTABLE}")
+  else()
+    find_program(DOCKER_EXECUTABLE docker)
+    if(DOCKER_EXECUTABLE)
+      add_ceph_test(run-promtool-unittests
+        {CMAKE_CURRENT_SOURCE_DIR}/run-promtool-unittests.sh "${DOCKER_EXECUTABLE}")
+    endif()
+  endif()
```
Comment (badone) [inline src/test/CMakeLists.txt:617]: We could similarly pull the prometheus binaries here if we didn't find docker or podman (although I guess this is unlikely)?

File: ceph.spec.in [hunk 4]
```diff
@@ -356,6 +356,7 @@ BuildRequires:	lz4-devel >= 1.7
 %if 0%{?fedora} || 0%{?rhel}
 BuildRequires:	libtool-ltdl-devel
 BuildRequires:	ninja-build
+BuildRequires:	podman
```
Comment (badone) [inline ceph.spec.in:359]: Leap and tumbleweed both have podman so maybe move this to a less restrictive section?

**Review by badone — CHANGES_REQUESTED** [review]
How do the libzbd and boost changes fit into this PR?

File: ceph.spec.in (hunk 4 above)
Comment (tchaikov) [reply ceph.spec.in:359]: I am going to drop this part. as I haven't tested this setting.

File: src/test/CMakeLists.txt (hunk 3 above)
Comment (tchaikov) [reply src/test/CMakeLists.txt:617]: I don't plan to pull the precompiled prometheus binaries for cater the needs for testing on distros which do not package prometheus or podman or docker .

File: src/test/CMakeLists.txt (hunk 2 above)
Comment (tchaikov) [reply src/test/CMakeLists.txt:604]: indeed, I don't want to pull the precompiled prometheus binary. and the goal of this PR is to address the testing on focal.

File: src/test/run-promtool-unittests.sh (hunk 1 above)
Comment (tchaikov) [reply src/test/run-promtool-unittests.sh:15]: okay, I will drop this change from this PR so this changeset can focus on the support on focal.

File: src/test/CMakeLists.txt (hunk 3 above)
Comment (badone) [reply src/test/CMakeLists.txt:617]: Good point.

File: src/test/run-promtool-unittests.sh (hunk 1 above)
Comment (badone) [reply src/test/run-promtool-unittests.sh:15]: You'd need to explain that comment because you give no context AFAICT.

Comment (tchaikov) [reply src/test/run-promtool-unittests.sh]: @badone would you kindly take another look?

**Review by badone — APPROVED** [review]

**Comment by tchaikov** [comment]
i included #40229 and #40230 to ensure that this changeset build on focal even these two changes are relatively unrelated to the goal of this changeset.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
```
The following tests FAILED:
	205 - unittest_interruptible_future (Failed)
	206 - unittest-seastar-errorator (Failed)
	207 - unittest-transaction-manager (Failed)
	209 - unittest-seastore-journal (Failed)
	210 - unittest-seastore-cache (Failed)
	211 - unittest-extmap-manager (Failed)
	212 - unittest-collection-manager (Failed)
	213 - unittest-omap-manager (Failed)
	214 - unittest-seastore (Failed)
	215 - unittest-staged-fltree (Failed)
	216 - unittest-fltree-onode-manager (Failed)
```

the test failures were due to the the number of jobs was `$(nproc) / 2`.  and on braggi18, we have 48 cores. so the crimson tests were competing with other tests for the file handle -- each reactor needs to use 1024 fd by default. i was watching
```console
$ watch -n1 cat /proc/sys/fs/aio-nr
```
the peak value was 3190848.

**Comment by tchaikov** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
changelog

- rebased against master.

**Review by avanthakkar — APPROVED** [review]

**Review by aaSharma14 — APPROVED** [review]

**Comment by tchaikov** [comment]
https://shaman.ceph.com/builds/ceph/wip-kefu2-testing-2021-03-19-2103/29bc5e2e91a5ed978465aabac7e990988e13db76/

**Pull request merged**